taichi/backends/opengl/struct_opengl.cpp

```diff
@@ -32,41 +31,41 @@ void OpenglStructCompiler::collect_snodes(SNode &snode) {
 
 void OpenglStructCompiler::generate_types(const SNode &snode) {
   const bool is_place = snode.is_place();
+  const auto &node_name = snode.node_type_name;
   if (!is_place) {
-    const std::string class_name = snode.node_type_name + "_ch";
+    const std::string child_name = node_name + "_ch";
     size_t stride_num = 0;
-    auto &gimme = class_get_map_[snode.node_type_name];
-    gimme.resize(snode.ch.size());
-    std::vector<std::pair<int, SNode *>> smp;
+    auto &snode_info = snode_map_[node_name];
+    snode_info.children_offsets.resize(snode.ch.size());
+    std::vector<std::pair<int, SNode *>> table;
     for (int i = 0; i < (int)snode.ch.size(); i++) {
-      smp.push_back(std::pair<int, SNode *>(i, snode.ch[i].get()));
+      table.push_back(std::pair<int, SNode *>(i, snode.ch[i].get()));
     }
     // Discussion: https://github.com/taichi-dev/taichi/issues/804
     // The NVIDIA OpenGL seems to stuck at compile (a very long time to compile)
     // when a small SNode is placed at a very high address... And I've no idea
     // why. So let's sort by SNode size, smaller first:
-    std::sort(smp.begin(), smp.end(),
+    std::sort(table.begin(), table.end(),
               [this](const std::pair<int, SNode *> &a,
                      const std::pair<int, SNode *> &b) {
-                return stride_map_.at(a.second->node_type_name) <
-                       stride_map_.at(b.second->node_type_name);
+                return snode_map_[a.second->node_type_name].stride <
+                       snode_map_[b.second->node_type_name].stride;
               });
-    for (auto &&[i, ch] : smp) {
-      gimme[i] = stride_num;
-      stride_num += stride_map_.at(ch->node_type_name);
+    for (auto &&[i, ch] : table) {
+      snode_info.children_offsets[i] = stride_num;
+      stride_num += snode_map_[ch->node_type_name].stride;
     }
-    stride_map_[class_name] = stride_num;
+    snode_map_[child_name].stride = stride_num;
   }
-  const auto &node_name = snode.node_type_name;
   if (is_place) {
     const auto dt_name = opengl_data_type_name(snode.dt);
-    stride_map_[node_name] = data_type_size(snode.dt);
+    snode_map_[node_name].stride = data_type_size(snode.dt);
   } else if (snode.type == SNodeType::dense || snode.type == SNodeType::root) {
-    const int n = (snode.type == SNodeType::dense) ? snode.n : 1;
-    length_map_[node_name] = n;
-    stride_map_[node_name] =
-        stride_map_[node_name + "_ch"] * length_map_[node_name];
-    class_children_map_[node_name] = stride_map_.at(node_name + "_ch");
+    const int n = (snode.type == SNodeType::root) ? 1 : snode.n;
+    snode_map_[node_name].length = n;
```

Comment:
Similar to what you've done for L38, you can write `auto &snode_info = snode_map_[node_name];` once?